File: src/bench/telemetry.py
"""Telemetry: what a run says to whoever is listening - its log records and its timed spans.

Both are the standards a collector already reads, so an OpenTelemetry exporter, Datadog or
Grafana can be attached by a host without anything here naming one.

**Logs** are the standard library's :mod:`logging`: one logger per module, under ``bench``.
Importing the package configures nothing - no handler, no level - because a library's records
are the host's to route: the app's worker attaches a handler to ``bench``, and a host that
attaches none gets :mod:`logging`'s own default, warnings on stderr. Structured fields ride
in ``extra`` under the one key :data:`FIELDS`; :func:`fields` builds that.

**Spans** are timed stretches of work, named the OpenTelemetry way - lowercase and dotted,
``bench.`` first - with attributes to match: ``bench.part.ref``, ``bench.mesh.triangles``, and
``error.type`` on a span whose work raised. A :class:`Tracer` is handed in at the edge the way a
kernel is, and :data:`SILENT`, which keeps nothing, is the default - so a run nobody is timing
pays for two clock reads a span and nothing else.
"""

import time
from collections.abc import Iterator, Mapping
from contextlib import contextmanager
from dataclasses import dataclass
from typing import NamedTuple, Protocol

Attribute = str | int | float | bool
"""What a span attribute or a log field may hold: the scalar kinds OpenTelemetry carries."""

FIELDS = "bench"
"""The ``extra`` key a log record's structured fields are kept under."""


class Span(NamedTuple):
    """One stretch of work: what it was, when it began (seconds since the epoch), how long it
    took (seconds), and what is worth knowing about it."""

    name: str
    start: float
    duration: float
    attributes: Mapping[str, Attribute]


class Tracer(Protocol):
    """Where finished spans go."""

    def record(self, span: Span, /) -> None: ...


@dataclass(frozen=True, slots=True)
class _Silent:
    """The tracer that keeps nothing."""

    def record(self, span: Span, /) -> None:
        return None


SILENT: Tracer = _Silent()
"""The default tracer: timing is on only when a host hands in one of its own."""


def fields(**values: Attribute) -> dict[str, dict[str, Attribute]]:
    """Structured fields for a log record, as ``extra`` carries them:
    ``log.info("built", extra=fields(parts=3))``."""
    return {FIELDS: values}


@contextmanager
def timed(tracer: Tracer, name: str, **attributes: Attribute) -> Iterator[dict[str, Attribute]]:
    """Time the block as one span called ``name``.

    The block is handed the span's attributes to add to - a count is often known only at the
    end. A block that raises is recorded all the same, with ``error.type`` naming what it
    raised, and the exception goes on its way untouched.
    """
    found: dict[str, Attribute] = dict(attributes)
    start = time.time()
    began = time.perf_counter()
    try:
        yield found
    except BaseException as exc:
        found["error.type"] = type(exc).__name__
        raise
    finally:
        tracer.record(Span(name, start, time.perf_counter() - began, found))

File: web/src/telemetry.ts
/** Telemetry: the app's log records and timed spans, and the one place a collector plugs in.
 *
 * Everything the app says about itself - the page, the worker, and the Python running in the
 * worker - arrives here as a `LogRecord` or a `SpanRecord` and goes to every attached `Sink`.
 * Two sinks ship: `consoleSink`, levelled, for a person with DevTools open, and
 * `userTimingSink`, which puts every span on this page's User Timing timeline as a
 * `performance.measure` - where the Performance panel draws it and where a RUM agent or
 * OpenTelemetry's web instrumentation already looks. Attaching Datadog, Grafana Faro or an
 * OTLP exporter is one more `attach` in `main.ts`; nothing else in the app changes.
 *
 * Names follow OpenTelemetry's conventions: spans are lowercase and dotted, `bench.` first,
 * with attributes such as `bench.part.ref`, and `error.type` on a span whose work threw.
 * Times are milliseconds since the epoch, so a span timed in the worker lands in the right
 * place on the page's timeline whatever the two threads' own time origins are.
 */

export type Attribute = string | number | boolean;
export type Attributes = Readonly<Record<string, Attribute>>;
export type Level = "debug" | "info" | "warn" | "error";

export interface LogRecord {
  readonly time: number;
  readonly level: Level;
  readonly logger: string;
  readonly message: string;
  readonly attributes: Attributes;
}

export interface SpanRecord {
  readonly name: string;
  readonly start: number;
  readonly duration: number;
  readonly attributes: Attributes;
}

/** Somewhere records go. Either half may be left out. */
export interface Sink {
  log?(record: LogRecord): void;
  span?(record: SpanRecord): void;
}

const RANK: Readonly<Record<Level, number>> = { debug: 10, info: 20, warn: 30, error: 40 };

/** Whether `value` names a level. */
export const isLevel = (value: unknown): value is Level =>
  typeof value === "string" && Object.hasOwn(RANK, value);

/** Milliseconds since the epoch, to the precision this thread's clock has. */
export const now = (): number => performance.timeOrigin + performance.now();

const sinks = new Set<Sink>();

/** Send every record from now on to `sink` as well. Returns the way to stop. */
export function attach(sink: Sink): () => void {
  sinks.add(sink);
  return () => {
    sinks.delete(sink);
  };
}

/** Hand one log record to every sink. A sink that throws is its own problem, not the app's. */
export function logged(record: LogRecord): void {
  for (const sink of sinks) {
    try {
      sink.log?.(record);
    } catch {
      // a broken sink must not take the thing it was watching down with it
    }
  }
}

/** Hand one finished span to every sink. */
export function spanned(record: SpanRecord): void {
  for (const sink of sinks) {
    try {
      sink.span?.(record);
    } catch {
      // as above
    }
  }
}

/** A log record stamped now. */
export function log(level: Level, logger: string, message: string, attributes: Attributes = {}): void {
  logged({ time: now(), level, logger, message, attributes });
}

/** Run `work` as one span called `name`, recorded whether it returns or throws. */
export function timed<T>(name: string, work: () => T, attributes: Attributes = {}): T {
  const start = now();
  let value: T;
  try {
    value = work();
  } catch (problem) {
    spanned({ name, start, duration: now() - start, attributes: withError(attributes, problem) });
    throw problem;
  }
  spanned({ name, start, duration: now() - start, attributes });
  return value;
}

/** The same for work that finishes later. */
export async function timedAsync<T>(
  name: string,
  work: () => Promise<T>,
  attributes: Attributes = {},
): Promise<T> {
  const start = now();
  try {
    const value = await work();
    spanned({ name, start, duration: now() - start, attributes });
    return value;
  } catch (problem) {
    spanned({ name, start, duration: now() - start, attributes: withError(attributes, problem) });
    throw problem;
  }
}

const withError = (attributes: Attributes, problem: unknown): Attributes => ({
  ...attributes,
  "error.type": problem instanceof Error ? problem.name : typeof problem,
});

/** The parts of `console` a sink writes to. */
export type ConsoleLike = Pick<Console, "debug" | "info" | "warn" | "error">;

/** Log records at `least` and above, to the console, with their fields beside the message. */
export function consoleSink(least: Level = "info", out: ConsoleLike = console): Sink {
  return {
    log(record) {
      if (RANK[record.level] < RANK[least]) return;
      out[record.level](`[${record.logger}] ${record.message}`, record.attributes);
    },
  };
}

/** Every span as a User Timing measure on this thread's timeline, its attributes as the
 * measure's `detail`. */
export const userTimingSink: Sink = {
  span(record) {
    try {
      performance.measure(record.name, {
        start: Math.max(0, record.start - performance.timeOrigin),
        duration: record.duration,
        detail: record.attributes,
      });
    } catch {
      // a timeline that will not take the entry - an old browser - loses the span, not the app
    }
  },
};
